### PR TITLE
dynare: import from homebrew/science

### PR DESCRIPTION
**Pull request by houtanb**
import from homebrew-science

Dynare is a software platform for handling a wide class of economic models, in particular dynamic stochastic general equilibrium (DSGE) and overlapping generations (OLG) models. The models solved by Dynare include those relying on the rational expectations hypothesis, wherein agents form their expectations about the future in a way consistent with the model. But Dynare is also able to handle models where expectations are formed differently: on one extreme, models where agents perfectly anticipate the future; on the other extreme, models where agents have limited rationality or imperfect knowledge of the state of  the economy and, hence, form their expectations through a learning process. In terms of types of agents, models solved  by Dynare can incorporate consumers, productive firms, governments, monetary authorities,  investors and financial intermediaries. Some degree of heterogeneity can be achieved by including several distinct classes of agents in each of the aforementioned agent categories.

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by houtanb** [comment]
This breaks because `libgfortran.a` is not found.....

**Comment by ilovezfs** [comment]
If we have this in homebrew/core it will need to be `octave` only without the matlab options.

**Comment by houtanb** [comment]
@ilovezfs why is this?

**Comment by ilovezfs** [comment]
Two primary reasons:
1. We've not been allowing options on imports from science (or the vast majority of new formulae for that matter) as they trigger source builds and consequent build failures and user issue reports.
2. There's no MATLAB available in Cask and Homebrew maintainers are unlikely to have access to a copy to be able to build, test, or debug that option.

**Comment by houtanb** [comment]
@ilovezfs thanks for the explanation. I'll modify the script to remove Matlab support.

File: Formula/dynare.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+class Dynare < Formula
+  desc "Platform for economic models, particularly DSGE and OLG models"
+  homepage "https://www.dynare.org/"
+  url "https://www.dynare.org/release/source/dynare-4.5.3.tar.xz"
+  sha256 "01434f6d3ceaff1891dc771f6a5b39caee787b2ffa1875e5a4e8c673e32ff3d7"
+  head "https://github.com/DynareTeam/dynare.git"
```
Comment (ilovezfs) [inline Formula/dynare.rb]: let's also remove head

File: Formula/dynare.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  depends_on "xz" => :build
+  depends_on "fftw"
+  depends_on "gcc" # for gfortran
+  depends_on "gsl"
+  depends_on "hdf5"
+  depends_on "libmatio"
+  depends_on "metis"
+  depends_on "readline"
+  depends_on "veclibfort"
+  depends_on "octave"
```
Comment (ilovezfs) [inline Formula/dynare.rb]: alphabetize the non-build deps amongst themselves

File: Formula/dynare.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  resource "slicot" do
+    url "https://mirrors.ocf.berkeley.edu/debian/pool/main/s/slicot/slicot_5.0+20101122.orig.tar.gz"
+    mirror "https://mirrorservice.org/sites/ftp.debian.org/debian/pool/main/s/slicot/slicot_5.0+20101122.orig.tar.gz"
+    sha256 "fa80f7c75dab6bfaca93c3b374c774fd87876f34fba969af9133eeaea5f39a3d"
+  end
+
+  def install
+    ENV.cxx11
+
+    args = %W[
```
Comment (ilovezfs) [inline Formula/dynare.rb]: these should all be inlined where you have *args instead of using an array

File: Formula/dynare.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    system "./configure", *args
+
+    system "make"
+    system "make", "install"
+  end
+
+  def caveats; <<~EOS
+    To get started with dynare, open Octave and type:
+
+            addpath #{opt_prefix}/lib/dynare/matlab
```
Comment (ilovezfs) [inline Formula/dynare.rb]: opt_lib

File: Formula/dynare.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    EOS
+  end
+
+  test do
+    cp lib/"dynare/examples/bkk.mod", testpath
+    if build.with? "octave"
```
Comment (ilovezfs) [inline Formula/dynare.rb]: the `if` should be removed

File: Formula/dynare.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      inreplace "m4/ax_mexopts.m4",
+        /MACOSX_DEPLOYMENT_TARGET='.*'/,
+        "MACOSX_DEPLOYMENT_TARGET='#{MacOS.version}'"
+
+      system "autoreconf", "-fvi"
+    end
+
+    system "./configure", *args
+
+    system "make"
```
Comment (ilovezfs) [inline Formula/dynare.rb]: does it actually fail without the separate make step?

File: Formula/dynare.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  if build.head?
+    depends_on "autoconf" => :build
+    depends_on "automake" => :build
+    depends_on "bison" => :build
+    depends_on "flex" => :build
+  end
+
+  depends_on "boost" => :build
+  depends_on "cweb" => :build
+  depends_on "xz" => :build
```
Comment (ilovezfs) [inline Formula/dynare.rb]: this is handled via magic. doesn't need to be here

File: Formula/dynare.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  depends_on "fftw"
+  depends_on "gcc" # for gfortran
+  depends_on "gsl"
+  depends_on "hdf5"
+  depends_on "libmatio"
+  depends_on "metis"
+  depends_on "readline"
```
Comment (ilovezfs) [inline Formula/dynare.rb]: BSD's libhistory is no good?

File: Formula/dynare.rb (hunk 1 above)
Comment (houtanb) [reply Formula/dynare.rb]: @ilovezfs is there any way to keep `HEAD`? It's a big reason for us to distribute with Homebrew....

**Comment by houtanb** [comment]
@ilovezfs I made all the changes you requested except for the removal of the `HEAD` install which is quite useful for our users. If there's any way to keep that in, that'd be very helpful to our community.

**Comment by ilovezfs** [comment]
@houtanb yes that's fine.

**Comment by houtanb** [comment]
@ilovezfs thank you so much

File: Formula/dynare.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  homepage "https://www.dynare.org/"
+  url "https://www.dynare.org/release/source/dynare-4.5.3.tar.xz"
+  sha256 "01434f6d3ceaff1891dc771f6a5b39caee787b2ffa1875e5a4e8c673e32ff3d7"
+
+  head do
+    url "https://github.com/DynareTeam/dynare.git"
+
+    depends_on "autoconf" => :build
+    depends_on "automake" => :build
+    depends_on "bison" => :build
```
Comment (DomT4) [inline Formula/dynare.rb]: macOS ships both `bison` and `flex`; it's worth trying removing these dependencies locally and seeing if the `HEAD` build still works.

File: Formula/dynare.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  def install
+    ENV.cxx11
+
+    resource("slicot").stage do
+      system "make", "lib", "OPTS=-fPIC", "SLICOTLIB=../libslicot_pic.a",
+             "FORTRAN=gfortran", "LOADER=gfortran"
+      system "make", "clean"
+      system "make", "lib", "OPTS=-fPIC -fdefault-integer-8",
+             "FORTRAN=gfortran", "LOADER=gfortran",
+             "SLICOTLIB=../libslicot64_pic.a"
```
Comment (DomT4) [inline Formula/dynare.rb]: There's a bunch of trailing whitespace on this line.

**Comment by DomT4** [comment]
There's a legit build failure in there too, apparently:
```
clang -fPIC -g -O2 -Wall -Wno-parentheses -bundle -bundle_loader /usr/local/Cellar/octave/4.2.1_10/bin/octave-4.2.1 -L/usr/local/Cellar/octave/4.2.1_10/lib/octave/4.2.1 -L/usr/local/Cellar/octave/4.2.1_10/lib  -o mjdgges.mex mjdgges.o  -loctinterp -loctave  -L/usr/local/opt/veclibfort/lib -lvecLibFort  -L/usr/local/Cellar/fftw/3.3.7/lib -lfftw3_threads -lfftw3 -L/usr/local/Cellar/fftw/3.3.7/lib -lfftw3f_threads -lfftw3f -lm    -L/usr/local/Cellar/gcc/7.2.0_1/lib/gcc/7/gcc/x86_64-apple-darwin16.7.0/7.2.0 -L/usr/local/Cellar/gcc/7.2.0_1/lib/gcc/7/gcc/x86_64-apple-darwin16.7.0/7.2.0/../../.. -lgfortran -lquadmath -lm
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [mjdgges.mex] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1
```

File: Formula/dynare.rb (hunk 10 above)
Comment (DomT4) [reply Formula/dynare.rb]: screen shot 2018-02-04 at 08 48 36

Comment (DomT4) [reply Formula/dynare.rb]: And the audit itself:
```
dynare:
  * C: 45: col 46: Trailing whitespace detected.
```

Comment (ilovezfs) [reply Formula/dynare.rb]:
```
iMac-TMP:homebrew-core joe$ cat -vet Formula/dynare.rb |grep SLICOTLIB
      system "make", "lib", "OPTS=-fPIC", "SLICOTLIB=../libslicot_pic.a",$
             "SLICOTLIB=../libslicot64_pic.a"      $
```
there we go. needed to use `brew pull --clean`

**Comment by ilovezfs** [comment]
Oh how wonderful
```
iMac-TMP:homebrew-core joe$ strings /usr/local/Cellar//octave/4.2.1_10/bin/mkoctfile|grep gcc
 -L/usr/local/Cellar/gcc/7.2.0_1/lib/gcc/7/gcc/x86_64-apple-darwin15.6.0/7.2.0 -L/usr/local/Cellar/gcc/7.2.0_1/lib/gcc/7/gcc/x86_64-apple-darwin15.6.0/7.2.0/../../.. -lgfortran -lquadmath -lm
```

**Comment by DomT4** [comment]
Oh, wow. That is _quite something_.

File: Formula/dynare.rb (hunk 9 above)
Comment (ilovezfs) [reply Formula/dynare.rb]: dies without them

```
bison -o DynareBison.cc DynareBison.yy
DynareBison.yy:21.10-14: require bison 2.5, but have 2.3
bison -o DynareBison.cc DynareBison.yy
DynareBison.yy:21.10-14: require bison 2.5, but have 2.3
make[1]: *** [stack.hh] Error 63
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [DynareBison.hh] Error 63
DynareBison.yy:21.10-14: require bison 2.5, but have 2.3
make[1]: *** [position.hh] Error 63
DynareBison.yy:21.10-14: require bison 2.5, but have 2.3
make[1]: *** [location.hh] Error 63
DynareBison.yy:21.10-14: require bison 2.5, but have 2.3
make[1]: *** [DynareBison.cc] Error 63
make: *** [install-recursive] Error 1
```
and
```
flex -o DynareFlex.cc DynareFlex.ll
cp /usr/bin/../include/FlexLexer.h . || test -f ./FlexLexer.h
cp: /usr/bin/../include/FlexLexer.h: No such file or directory
make[1]: *** [FlexLexer.h] Error 1
make[1]: *** Waiting for unfinished jobs....
cp /usr/bin/../include/FlexLexer.h . || test -f ./FlexLexer.h
cp: /usr/bin/../include/FlexLexer.h: No such file or directory
make[1]: *** [DynareFlex.cc] Error 1
make: *** [install-recursive] Error 1
```
```

**Comment by ilovezfs** [comment]
@houtanb the test block fails for `--HEAD`
```
==> /usr/local/opt/octave/bin/octave --no-gui -H --path /usr/local/opt/dynare/lib/dynare/matlab --eval dynare bkk.mod console
warning: function /usr/local/opt/dynare/lib/dynare/matlab/missing/corrcoef/tcdf.m shadows a core library function
warning: called from
    dynare_config at line 142 column 1
    dynare at line 69 column 12

Configuring Dynare ...
[mex] Generalized QZ.
[mex] Sylvester equation solution.
[mex] Kronecker products.
[mex] Sparse kronecker products.
[mex] Local state space iteration (second order).
[mex] Bytecode evaluation.
[mex] k-order perturbation solver.
[mex] k-order solution simulation.
[mex] Quasi Monte-Carlo sequence (Sobol).
[mex] Markov Switching SBVAR.

error: urlwrite: unable to open file
/usr/local/Homebrew/Library/Homebrew/test.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python.rb
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
BuildError: Failed executing: /usr/local/opt/octave/bin/octave --no-gui -H --path /usr/local/opt/dynare/lib/dynare/matlab --eval dynare\ bkk.mod\ console
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 
```

File: Formula/dynare.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                          "--prefix=#{prefix}",
+                          "--disable-matlab",
+                          "--with-slicot=#{buildpath}/slicot"
+    system "make", "install"
+  end
+
+  def caveats; <<~EOS
+    To get started with dynare, open Octave and type:
+
+            addpath #{opt_lib}/dynare/matlab
```
Comment (DomT4) [inline Formula/dynare.rb]: @ilovezfs Plz fix this massive indentation before merging 🙈.

Comment (ilovezfs) [reply Formula/dynare.rb]: don't worry. nothing is getting merged until `--HEAD` passes the test block

Comment (DomT4) [reply Formula/dynare.rb]: Maybe we should add an `opt_pkglib` for `lib/name` 😈.

**Comment by ilovezfs** [comment]
@houtanb ping on the `--HEAD` test block failure.

**Comment by houtanb** [comment]
@ilovezfs yes, it's next on my list. I have something urgent to finish. As soon as that's done I'll turn to this. Sorry about that.

**Comment by ilovezfs** [comment]
@houtanb no worries

**Comment by houtanb** [comment]
@ilovezfs the only explanation I can provide for the failure in the `HEAD` test is that the submodules were either not downloaded or initialized. I don't see how this is possible as, having followed the output of `brew install --debug --verbose --HEAD dynare`, I see that they are indeed fetched and initialized. When I run this command locally, the installation proceeds without problem. Would it be possible to try again on your setup?

To go into more detail, we have one submodule, [`dseries`](https://github.com/DynareTeam/dseries) that depends on another submodule, [`dates`](https://github.com/DynareTeam/dates). If the later is not found, then we instruct Matlab/Octave to download it on line 22 of `modules/dseries/src/initialize_dseries_toolbox.m`:
```
modules/dseries/src/initialize_dseries_toolbox.m:22:    urlwrite('https://github.com/DynareTeam/dates/archive/master.zip','master.zip');
```
Other than this, we only use `urlwrite` for internal tests, none of which would be called when running the test command above

**Comment by houtanb** [comment]
@ilovezfs also note that I bumped the version of Dynare as a new version was released on Saturday.

**Comment by houtanb** [comment]
@ilovezfs just an update, the end of my local build process looks like this, so I don't see where the test is actually run....
```
Fixing /usr/local/Cellar/dynare/HEAD-4adebd4/lib/dynare/mex/octave/qzcomplex.oct permissions from 755 to 444
Fixing /usr/local/Cellar/dynare/HEAD-4adebd4/lib/dynare/mex/octave/sparse_hessian_times_B_kronecker_C.mex permissions from 755 to 444
rm /usr/local/Cellar/dynare/HEAD-4adebd4/share/info/dir
==> Finishing up
ln -s ../Cellar/dynare/HEAD-4adebd4/bin/dynare++ dynare++
ln -s ../../Cellar/dynare/HEAD-4adebd4/share/info/dynare.info dynare.info
info /usr/local/share/info/dynare.info
ln -s ../../Cellar/dynare/HEAD-4adebd4/share/info/dynare.info-1 dynare.info-1
ln -s ../../Cellar/dynare/HEAD-4adebd4/share/info/dynare.info-2 dynare.info-2
ln -s ../Cellar/dynare/HEAD-4adebd4/lib/dynare dynare
/usr/bin/sandbox-exec -f /Users/houtanb/scratch/homebrew_tmp/homebrew20180206-28139-18o3msg.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/postinstall.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/dynare.rb --debug --verbose --HEAD --HEAD
/usr/local/Homebrew/Library/Homebrew/postinstall.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/dynare.rb
==> Caveats
To get started with dynare, open Octave and type
  addpath /usr/local/opt/dynare/lib/dynare/matlab
==> Summary
🍺  /usr/local/Cellar/dynare/HEAD-4adebd4: 1,442 files, 20.8MB, built in 3 minutes 28 seconds
```

**Comment by ilovezfs** [comment]
You need to manually run `brew  test --verbose --debug dynare`

**Comment by houtanb** [comment]
@ilovezfs thanks. I see the problem. I'll push a fix and let you know when the test passes on `HEAD`

**Comment by ilovezfs** [comment]
Cool

**Comment by houtanb** [comment]
@ilovezfs the `HEAD` test works now.

**Review by ilovezfs — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @houtanb for the pull request!

**Comment by houtanb** [comment]
Thank you very much @ilovezfs and @fxcoudert

**Pull request closed without merging**